File: src/BaseHmiTypes/Screens/Base/HmiProperty.cs
namespace BaseHmiTypes.Screens.Base;

public enum HmiPropertyKind
{
    Static,
    Default,
    Tag,
    Script,
    Expression,
    Blink
}

public enum HmiScriptLanguage
{
    Unknown,
    C,
    VBScript,
    JavaScript
}

public enum HmiBlinkCondition
{
    Always,
    WhenTrue,
    WhenFalse
}

public enum HmiBlinkRate
{
    Default,
    Slow,
    Medium,
    Fast
}

public enum HmiTriggerKind
{
    Tag,
    Cycle,
    Event
}

public enum HmiTagTriggerMode
{
    ValueChange,
    RisingEdge,
    FallingEdge
}

public enum HmiValueConverterKind
{
    Unknown,
    Linear,
    Range,
    Expression,
    Script
}

public abstract class HmiProperty<T>
{
    public abstract HmiPropertyKind Kind { get; }

    public virtual T? StaticValue { get; set; }

    public static implicit operator HmiProperty<T>(T? value)
    {
        return new HmiStaticProperty<T> { StaticValue = value };
    }
}

public abstract class HmiDynamicProperty<T> : HmiProperty<T>
{
    public IList<HmiTrigger> Triggers { get; } = new List<HmiTrigger>();
}

public abstract class HmiTrigger
{
    public abstract HmiTriggerKind Kind { get; }

    public string? Name { get; set; }
}

public sealed class HmiTagTrigger : HmiTrigger
{
    public override HmiTriggerKind Kind => HmiTriggerKind.Tag;

    public IList<string> TagNames { get; } = new List<string>();

    public HmiTagTriggerMode Mode { get; set; } = HmiTagTriggerMode.ValueChange;
}

public sealed class HmiCycleTrigger : HmiTrigger
{
    public override HmiTriggerKind Kind => HmiTriggerKind.Cycle;

    public string? CycleName { get; set; }

    public int? CycleTime { get; set; }

    public string? CycleUnit { get; set; }
}

public sealed class HmiEventTrigger : HmiTrigger
{
    public override HmiTriggerKind Kind => HmiTriggerKind.Event;

    public string? EventName { get; set; }
}

public sealed class HmiValueConverter
{
    public HmiValueConverterKind Kind { get; set; } = HmiValueConverterKind.Unknown;

    public string? Name { get; set; }

    public string? Expression { get; set; }

    public HmiScriptLanguage Language { get; set; } = HmiScriptLanguage.Unknown;

    public string? Script { get; set; }

    public IDictionary<string, object?> Parameters { get; } = new Dictionary<string, object?>();
}

public sealed class HmiStaticProperty<T> : HmiProperty<T>
{
    public override HmiPropertyKind Kind => HmiPropertyKind.Static;
}

public sealed class HmiDefaultProperty<T> : HmiProperty<T>
{
    public override HmiPropertyKind Kind => HmiPropertyKind.Default;

    public string? ProfileName { get; set; }

    public string? PropertyName { get; set; }
}

public sealed class HmiTagProperty<T> : HmiDynamicProperty<T>
{
    public override HmiPropertyKind Kind => HmiPropertyKind.Tag;

    public string? TagName { get; set; }
}

public sealed class HmiScriptProperty<T> : HmiDynamicProperty<T>
{
    public override HmiPropertyKind Kind => HmiPropertyKind.Script;

    public HmiScriptLanguage Language { get; set; } = HmiScriptLanguage.Unknown;

    public string? Script { get; set; }
}

public sealed class HmiExpressionProperty<T> : HmiDynamicProperty<T>
{
    public override HmiPropertyKind Kind => HmiPropertyKind.Expression;

    public string? Expression { get; set; }

    public IList<HmiValueConverter> Converters { get; } = new List<HmiValueConverter>();
}

public sealed class HmiBlinkProperty<T> : HmiProperty<T>
{
    public override HmiPropertyKind Kind => HmiPropertyKind.Blink;

    public T? BlinkValue { get; set; }

    public HmiBlinkCondition Condition { get; set; } = HmiBlinkCondition.Always;

    public HmiBlinkRate Rate { get; set; } = HmiBlinkRate.Default;

    public string? ConditionTagName { get; set; }
}

public static class HmiProperty
{
    public static HmiStaticProperty<T> Static<T>(T? value)
    {
        return new HmiStaticProperty<T> { StaticValue = value };
    }

    public static HmiDefaultProperty<T> Default<T>(T? value, string? profileName = null, string? propertyName = null)
    {
        return new HmiDefaultProperty<T>
        {
            StaticValue = value,
            ProfileName = profileName,
            PropertyName = propertyName
        };
    }

    public static HmiTagProperty<T> Tag<T>(string tagName, T? fallbackValue = default)
    {
        return new HmiTagProperty<T> { TagName = tagName, StaticValue = fallbackValue };
    }

    public static HmiScriptProperty<T> Script<T>(string script, HmiScriptLanguage language, T? fallbackValue = default)
    {
        return new HmiScriptProperty<T> { Script = script, Language = language, StaticValue = fallbackValue };
    }

    public static HmiExpressionProperty<T> Expression<T>(string expression, T? fallbackValue = default)
    {
        return new HmiExpressionProperty<T> { Expression = expression, StaticValue = fallbackValue };
    }

    public static HmiBlinkProperty<T> Blink<T>(
        T? staticValue,
        T? blinkValue,
        HmiBlinkRate rate = HmiBlinkRate.Default,
        HmiBlinkCondition condition = HmiBlinkCondition.Always,
        string? conditionTagName = null)
    {
        return new HmiBlinkProperty<T>
        {
            StaticValue = staticValue,
            BlinkValue = blinkValue,
            Rate = rate,
            Condition = condition,
            ConditionTagName = conditionTagName
        };
    }
}

public static class HmiPropertyExtensions
{
    public static T? GetStaticValue<T>(this HmiProperty<T>? property)
    {
        return property == null ? default : property.StaticValue;
    }

    public static T GetStaticValueOrDefault<T>(this HmiProperty<T>? property, T defaultValue = default!)
    {
        return property == null ? defaultValue : property.StaticValue ?? defaultValue;
    }
}

File: typescript/base-hmi-types/src/screens/base/HmiProperty.ts
export enum HmiPropertyKind {
  Static = "Static",
  Tag = "Tag",
  Script = "Script",
  Expression = "Expression",
  Blink = "Blink",
}

export enum HmiScriptLanguage {
  Unknown = "Unknown",
  C = "C",
  VBScript = "VBScript",
  JavaScript = "JavaScript",
}

export enum HmiBlinkCondition {
  Always = "Always",
  WhenTrue = "WhenTrue",
  WhenFalse = "WhenFalse",
}

export enum HmiBlinkRate {
  Default = "Default",
  Slow = "Slow",
  Medium = "Medium",
  Fast = "Fast",
}

export enum HmiTriggerKind {
  Tag = "Tag",
  Cycle = "Cycle",
  Event = "Event",
}

export enum HmiTagTriggerMode {
  ValueChange = "ValueChange",
  RisingEdge = "RisingEdge",
  FallingEdge = "FallingEdge",
}

export enum HmiValueConverterKind {
  Unknown = "Unknown",
  Linear = "Linear",
  Range = "Range",
  Expression = "Expression",
  Script = "Script",
}

export interface HmiProperty<T> {
  readonly kind: HmiPropertyKind;
  staticValue?: T;
}

export interface HmiDynamicProperty<T> extends HmiProperty<T> {
  triggers: HmiTrigger[];
}

export interface HmiTrigger {
  readonly kind: HmiTriggerKind;
  name?: string;
}

export interface HmiTagTrigger extends HmiTrigger {
  readonly kind: HmiTriggerKind.Tag;
  tagNames: string[];
  mode: HmiTagTriggerMode;
}

export interface HmiCycleTrigger extends HmiTrigger {
  readonly kind: HmiTriggerKind.Cycle;
  cycleName?: string;
  cycleTime?: number;
  cycleUnit?: string;
}

export interface HmiEventTrigger extends HmiTrigger {
  readonly kind: HmiTriggerKind.Event;
  eventName?: string;
}

export interface HmiValueConverter {
  kind: HmiValueConverterKind;
  name?: string;
  expression?: string;
  language: HmiScriptLanguage;
  script?: string;
  parameters: Record<string, unknown>;
}

export interface HmiStaticProperty<T> extends HmiProperty<T> {
  readonly kind: HmiPropertyKind.Static;
}

export interface HmiTagProperty<T> extends HmiDynamicProperty<T> {
  readonly kind: HmiPropertyKind.Tag;
  tagName?: string;
}

export interface HmiScriptProperty<T> extends HmiDynamicProperty<T> {
  readonly kind: HmiPropertyKind.Script;
  language: HmiScriptLanguage;
  script?: string;
}

export interface HmiExpressionProperty<T> extends HmiDynamicProperty<T> {
  readonly kind: HmiPropertyKind.Expression;
  expression?: string;
  converters: HmiValueConverter[];
}

export interface HmiBlinkProperty<T> extends HmiProperty<T> {
  readonly kind: HmiPropertyKind.Blink;
  blinkValue?: T;
  condition: HmiBlinkCondition;
  rate: HmiBlinkRate;
  conditionTagName?: string;
}

export function staticProperty<T>(value?: T): HmiStaticProperty<T> {
  return { kind: HmiPropertyKind.Static, staticValue: value };
}

export function tagProperty<T>(tagName: string, fallbackValue?: T): HmiTagProperty<T> {
  return { kind: HmiPropertyKind.Tag, tagName, staticValue: fallbackValue, triggers: [] };
}

export function scriptProperty<T>(
  script: string,
  language: HmiScriptLanguage,
  fallbackValue?: T,
): HmiScriptProperty<T> {
  return { kind: HmiPropertyKind.Script, language, script, staticValue: fallbackValue, triggers: [] };
}

export function expressionProperty<T>(expression: string, fallbackValue?: T): HmiExpressionProperty<T> {
  return { kind: HmiPropertyKind.Expression, expression, staticValue: fallbackValue, triggers: [], converters: [] };
}

export function blinkProperty<T>(
  staticValue?: T,
  blinkValue?: T,
  rate: HmiBlinkRate = HmiBlinkRate.Default,
  condition: HmiBlinkCondition = HmiBlinkCondition.Always,
  conditionTagName?: string,
): HmiBlinkProperty<T> {
  return { kind: HmiPropertyKind.Blink, staticValue, blinkValue, rate, condition, conditionTagName };
}

export function getStaticValue<T>(property?: HmiProperty<T>): T | undefined {
  return property?.staticValue;
}

export function getStaticValueOrDefault<T>(property: HmiProperty<T> | undefined, defaultValue: T): T {
  return property?.staticValue ?? defaultValue;
}

export class HmiPropertyExtensions {
  static getStaticValue<T>(property?: HmiProperty<T>): T | undefined {
    return getStaticValue(property);
  }

  static getStaticValueOrDefault<T>(property: HmiProperty<T> | undefined, defaultValue: T): T {
    return getStaticValueOrDefault(property, defaultValue);
  }
}
